### ISSUE TITLE
some question about implementing gradient normalization clip

### ISSUE THREAD

**Issue by moveforever** [apache/incubator-mxnet#5062]
i implement  gradient normalization clip as followed in file optimizer.py
```
        norms = norm(grad)//mx.nd.norm
        if  norms > self.gradient_norm_threshold:
            grad = grad * self.gradient_norm_threshold/norms
```
is it right?

**Comment by feiyulv** [comment]
You can refer to this #4177

**Comment by yudeshui** [comment]
Is this implement is in update() in optimizer.py, and access only one parameter? @moveforever

**Comment by yudeshui** [comment]
@feiyulv I have access the issue #4177, and know the implementation of the standard gradient clip with all parameters. 
Do you solve the problem about how to get the parameters list (like the 'params_grad' in function 'norm_clipping') for we can only access one parameter in the mxnet's optimizer.
Thank a lot :)

**Comment by moveforever** [comment]
@yudeshui , yes, there only one parameter which is scalar, vector or matrix in function update. 
if as follow #4177, you have to reimplement all the process including load batch data, updating parameter.

**Comment by yudeshui** [comment]
@moveforever  Thanks :). And did you implement the idea in #4177 ?

**Comment by sxjscience** [comment]
What I'm using now is like this
```python
class MyModule(Module):
    """Some enhancement to the mx.mod.Module

    """
    def norm_clipping(self, threshold=1.0):
        """Clip the norm according to the threshold.
        All the gradients are concatenated to a single vector and the overall norm is calculated.
        Follows `[ICML2013] On the difficulty of training recurrent neural networks`

        Parameters
        ----------
        threshold : float, optional

        Returns
        -------
        norm_val : float
            The norm value. It could be used to measure whether the gradients are stable.
        """
        assert self.binded and self.params_initialized
        norm_val = self.get_global_norm_val()
        if norm_val > threshold:
            ratio = threshold / float(norm_val)
            for grads in self._exec_group.grad_arrays:
                for grad in grads:
                    grad[:] *= ratio
        return norm_val

    def get_global_norm_val(self):
        """Get the overall gradient norm ||W||_2

        Parameters
        ----------
        net : mx.mod.Module

        Returns
        -------
        norm_val : float
        """
        assert self.binded and self.params_initialized
        #TODO The code in the following will cause the estimated norm to be different for multiple gpus
        norm_val = 0.0
        for i in range(len(self._exec_group.grad_arrays[0])):
            norm_val += np.sqrt(
                sum([nd.norm(grads[i]).asnumpy()[0] ** 2
                     for grads in self._exec_group.grad_arrays]))
        norm_val /= float(len(self._exec_group.grad_arrays[0]))
        return norm_val
```
The training workflow is like:
```python
net.forward()
net.backward()
net.norm_clipping(threshold=...)
net.update()
```
This feature is still not ready for PR due to the multi-GPU issue. The implementation will give incorrect norm for multi-gpu training.

**Comment by yudeshui** [comment]
@sxjscience  It's very helpful for me. Thanks a lot. :)

**Comment by sxjscience** [comment]
@yudeshui Could you confirm that this works in your case? If that is so, I can PR this first and doc the incorrect behavior for multi-GPU training.

**Comment by yudeshui** [comment]
@sxjscience Yes. I use the implement in my work with single GPU and it works well.

**Comment by moveforever** [comment]
@sxjscience 
i imitate your implementation to support multi-gpu. And i use mx.nd operation as soon as possible. The code is as followed:
`
    def norm_clipping(self, threshold=1.0):
        """Clip the norm according to the threshold.
        All the gradients are concatenated to a single vector and the overall norm is calculated.
        Follows `[ICML2013] On the difficulty of training recurrent neural networks`

        Parameters
        ----------
        threshold : float, optional

        Returns
        -------
        norm_val : float
            The norm value. It could be used to measure whether the gradients are stable.
        """
        assert self.binded and self.params_initialized
        norm_val = self.get_global_norm_val()
	for i in range(len(self._exec_group.execs)):
            if norm_val[i] > threshold:
           	ratio = threshold / float(norm_val)
            	for grads in self._exec_group.grad_arrays:
                    grads[i][:] *= ratio

    def get_global_norm_val(self):
        """Get the overall gradient norm ||W||_2

        Parameters
        ----------
        net : mx.mod.Module

        Returns
        -------
        norm_val : float
        """
        assert self.binded and self.params_initialized
	import numpy as np
        norm_val = np.zeros(len(self._exec_group.execs))
	for grad_arr in self._exec_group.grad_arrays:
	    for i in range(len(self._exec_group.execs)):
		l2_norm = nd.square(nd.norm(grad_arr[i] / self._exec_group.batch_size)).asscalar()
		norm_val[i] += l2_norm
        return norm_val`

**Comment by moveforever** [comment]
@sxjscience 
i imitate your implementation to support multi-gpu. And i use mx.nd operation as soon as possible. 
i am not sure whether it's right
The code is as followed:

`    def norm_clipping(self, threshold=1.0):
        """Clip the norm according to the threshold.
        All the gradients are concatenated to a single vector and the overall norm is calculated.
        Follows `[ICML2013] On the difficulty of training recurrent neural networks`

        Parameters
        ----------
        threshold : float, optional

        Returns
        -------
        norm_val : float
            The norm value. It could be used to measure whether the gradients are stable.
        """
        assert self.binded and self.params_initialized
        norm_val = self.get_global_norm_val()
	for i in range(len(self._exec_group.execs)):
            if norm_val[i] > threshold:
           	ratio = threshold / float(norm_val[i])
            	for grads in self._exec_group.grad_arrays:
                    grads[i][:] *= ratio

    def get_global_norm_val(self):
        """Get the overall gradient norm ||W||_2

        Parameters
        ----------
        net : mx.mod.Module

        Returns
        -------
        norm_val : float
        """
        assert self.binded and self.params_initialized
        #TODO The code in the following will cause the estimated norm to be different for multiple gpus
	import numpy as np
        norm_val = np.zeros(len(self._exec_group.execs))
	for grad_arr in self._exec_group.grad_arrays:
	    for i in range(len(self._exec_group.execs)):
		l2_norm = nd.square(nd.norm(grad_arr[i] / self._exec_group.batch_size)).asscalar()
		norm_val[i] += l2_norm
        return np.sqrt(norm_val)
`

**Comment by sxjscience** [comment]
Thanks very much! I'll go back to Hong Kong tomorrow and will begin to finish this.

Xingjian

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by sxjscience** [comment]
We have gluon now and all the implementations should be shifted to gluon.
